webui-src/app/files/files_downloads.js

```diff
@@ -84,21 +84,26 @@ function addFile(url) {
   }
   rs.rsJsonApiRequest(
     '/rsFiles/FileRequest',
-    {
-      fileName: details.name,
-      hash: details.hash,
-
-      flags: util.RS_FILE_REQ_ANONYMOUS_ROUTING,
-      size: details.size,
-    },
+    {},
     (status) => {
       widget.popupMessage([
         m('i.fas.fa-file-medical'),
         m('h3', 'Add new file'),
         m('hr'),
         m('p', 'Successfully added file!'),
       ]);
-    }
+    },
+    true,
+    {},
+    undefined,
+    // NOTE: Payload is passed through serializer since JS cannot
+    // represent filesizes >2Gi using `Number` accurately
+    () => `{
+        "fileName": "${details.name}",
+        "hash": "${details.hash}",
+        "flags": ${util.RS_FILE_REQ_ANONYMOUS_ROUTING},
+        "size": ${details.size}
```

Comment:
I have on doubt  `details.size ` is a already a string not a number. without serializer It should also send the file size correct.

Reply:
Default serializer would send the size as a string (with the `""` quotes). Here the size is written to json as a number.

Reply:
If you look at this code: https://github.com/RetroShare/RSNewWebUI/blob/a4748df0dda7698b3b9bd907dd790673d34342d9/webui-src/app/files/files_downloads.js#L76

`details.size`  is extracted from the  retroshare file link. So the datatype of details size should be string not a number

Reply:
if I understand correctly how it is done for lobby_id in the webUI code (which is a 64 bits number as well), RS expects the xstr64 field to be filled by the webUI and converts it itself to a number, if not it uses the xint64 (default) which contains a 64bits number.

Reply:
The idea being that the webUI should never handle 64bit ints as numbers, but always keep them as strings, and pass them to RS using that string value xstr64 instead of xint64, so that RS does the conversion itself.

Reply:
> if I understand correctly how it is done for lobby_id in the webUI code (which is a 64 bits number as well), RS expects the xstr64 field to be filled by the webUI and converts it itself to a number, if not it uses the xint64 (default) which contains a 64bits number.


yes i looked at that codebase. they also  send `xstr64` in the place of `xint64`. That's  why i also send the string instead of number in `FileRequest` JSON API call. According to me, it should work.

Reply:
Yes, it's a string. By using the custom serializer, I'm simply avoiding the quotes from the data, so that the size is recognized as a number.
For example:
Old:
```json
{ "size": "12345" }
```
New:
```json
{ "size": 12345 }
```

Reply:
Thanks @rottencandy. I donot know about the unexpected behaviour of the custom serializer.

Reply:
Does this work? It doesn't seem to work for me.

Reply:
You need to use this:

        "size": { "xstr64": "${details.size}" }

then it works!

Reply:
I let you do the change, I'll merge the PR afterwards

Reply:
Hmm, not sure why the number method doesn't work.
But I've updated the call to use xstr64. I guess we won't be needing custom serialization in this case.

Reply:
I think that if you send as a number, js first converts ${details.size} to a double, then sends to RS as a number => wrong result. Maybe "${details.size}" would just work? I haven't tried.